### DISCUSSION TITLE
Queue not work as exected

### DISCUSSION THREAD

**Discussion by moppi79** [micropython/micropython#15953] [Using MicroPython] [answered]
hi 

i get stuck with queue
i want the data from the server, in the Main loop. 
But wenn i call the queue, i get back a  <generator object>

server on the ESP , working fast and fine. 

ESP32 (Wroom) code (running on micropython v1.23.0)

```py

import time, asyncio
from primitives import Queue

class server_task():
	
	def __init__ (self):
		self.bbb = 'test text'
		self.q = Queue()

	async def test(self,reader, writer):
		print (self.bbb,'main')

		print ('server start')
		data = await reader.read(100)
		print ('data IN:', data.decode())
		await self.q.put(data.decode())
		a = str('i am batman, not a return-server').encode()
		writer.write(a)
		await writer.drain()

		writer.close()
		await writer.wait_closed()


	async def run_server(self):
		server = await asyncio.start_server(self.test,'192.168.1.160',18888)
		async with server: await server.task
		
	def get (self):
		print ('here get')
		ret = self.q.get()
		print ('get Queue: ',ret)
		return('return data')


async def main():
	cc = server_task()
	asyncio.create_task(cc.run_server())
	while True:
		await asyncio.sleep(0.5)
		h = cc.get()
		print ('return get_in_loop: ',h)
		print ('loop')

asyncio.run(main())
```

ESP CLI output 

> ...
here get
get Queue:  <generator object 'get' at 3ffda0f0>
return get_in_loop:  return data
loop
test text main
server start
data IN: Hello World!
here get
get Queue:  <generator object 'get' at 3ffda430>
return get_in_loop:  return data
loop
here get
...

can someone tell me where my mistake is
i have no problem to call or send data to my a other Python-server (this works with queue too, but with threads).
when it is need i can post my client code as well 

BR 
Moppi

**Comment by bidrohini** [comment, marked as answer]
You can edit your code like this:
import time, asyncio
from primitives import Queue

class server_task:
    def __init__(self):
        self.bbb = 'test text'
        self.q = Queue()

    async def test(self, reader, writer):
        print(self.bbb, 'main')

        print('server start')
        data = await reader.read(100)
        print('data IN:', data.decode())
        await self.q.put(data.decode())  # Put received data into the queue
        a = str('i am batman, not a return-server').encode()
        writer.write(a)
        await writer.drain()

        writer.close()
        await writer.wait_closed()

    async def run_server(self):
        server = await asyncio.start_server(self.test, '192.168.1.160', 18888)
        async with server:
            await server.serve_forever()

    async def get(self):
        print('here get')
        ret = await self.q.get()  # Await the queue to get the data
        print('get Queue: ', ret)
        return ret  # Return the actual data from the queue

async def main():
    cc = server_task()
    asyncio.create_task(cc.run_server())
    while True:
        await asyncio.sleep(0.5)
        h = await cc.get()  # Await the get method to retrieve data
        print('return get_in_loop: ', h)
        print('loop')


asyncio.run(main())

The get() method in the server_task class is now an async def, and it awaits the self.q.get() method to fetch data from the queue properly.
In the main() function, when calling cc.get(), it now awaits the result to handle the asynchronous queue correctly.

**Comment by moppi79** [comment]
Thanks a Lot ! :)

it finally working :D 

i have in main loop put a wait_for so the loop is steady running 

```py
try:
	h = await asyncio.wait_for(cc.get(), timeout=0.5)
except :
	print('timeout!')
```

BR 
Moppi
